File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    que = deque()
    que.append([0,0,1])
    dy, dx = [0,0,1,-1], [1,-1,0,0]
    n = len(maps)
    m = len(maps[0])
    visit = [[False] * m for _ in range(n)]
    visit[0][0] = True
    
    while que:
        cur_y, cur_x, cur_cnt = que.popleft()
        
        for i in range(4):
            next_y, next_x, next_cnt = cur_y + dy[i], cur_x + dx[i], cur_cnt + 1
            
            if 0 <= next_x < m and 0 <= next_y < n and\
            maps[next_y][next_x] and not visit[next_y][next_x]:
                if next_y == n-1 and next_x == m-1:
                    return next_cnt
                visit[next_y][next_x] = True
                que.append([next_y, next_x, next_cnt])
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        ArrayDeque<int[]> que = new ArrayDeque<>();
        que.add(new int[]{0, 0, 1});
        int n = maps.length, m = maps[0].length;
        
        boolean[][] visit = new boolean[n][m];
        visit[0][0] = true;
        
        int[] dy = {0,0,1,-1}, dx = {-1,1,0,0};
        while(!que.isEmpty()){
            int[] cur = que.poll();
            
            int y = cur[0], x = cur[1];
            for(int i=0 ; i<4 ; i++){
                int ny = y + dy[i], nx = x + dx[i];
                
                if(0 > ny || ny >= n || 0 > nx || nx >= m || visit[ny][nx] || maps[ny][nx] == 0)
                    continue;
                
                if(ny == n - 1 && nx == m - 1)
                    return cur[2] + 1;
                
                que.add(new int[]{ny, nx, cur[2] + 1});
                visit[ny][nx] = true;
            }
            
        }
        return -1;
    }
}